C33 uses the asm-generic IPC layouts: time64 already supplies both
32-bit timestamp words. Avoid extra legacy padding in shm and sem.
Only translate timestamps after successful stat commands, never into
IPC_INFO buffers or semctl integer arguments; read varargs only when needed.

diff --git a/libc/sysdeps/linux/common/bits/shm.h b/libc/sysdeps/linux/common/bits/shm.h
--- a/libc/sysdeps/linux/common/bits/shm.h
+++ b/libc/sysdeps/linux/common/bits/shm.h
@@ -55,8 +55,8 @@
 #else
     __time_t shm_atime;			/* time of last shmat() */
 #endif
-#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__m68k__) && !defined(__microblaze__)) || \
-    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__m68k__) || defined(__microblaze__)) && !defined(__UCLIBC_USE_TIME64__))
+#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__m68k__) && !defined(__c33__) && !defined(__microblaze__)) || \
+    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__m68k__) || defined(__c33__) || defined(__microblaze__)) && !defined(__UCLIBC_USE_TIME64__))
     unsigned long int __uclibc_unused1;
 #endif
 #if defined(__UCLIBC_USE_TIME64__)
@@ -65,8 +65,8 @@
 #else
     __time_t shm_dtime;			/* time of last shmdt() */
 #endif
-#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__m68k__) && !defined(__microblaze__)) || \
-    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__m68k__) || defined(__microblaze__)) && !defined(__UCLIBC_USE_TIME64__))
+#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__m68k__) && !defined(__c33__) && !defined(__microblaze__)) || \
+    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__m68k__) || defined(__c33__) || defined(__microblaze__)) && !defined(__UCLIBC_USE_TIME64__))
     unsigned long int __uclibc_unused2;
 #endif
 #if defined(__UCLIBC_USE_TIME64__)
@@ -75,8 +75,8 @@
 #else
     __time_t shm_ctime;			/* time of last change by shmctl() */
 #endif
-#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__m68k__) && !defined(__microblaze__)) || \
-    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__m68k__) || defined(__microblaze__)) && !defined(__UCLIBC_USE_TIME64__))
+#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__m68k__) && !defined(__c33__) && !defined(__microblaze__)) || \
+    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__m68k__) || defined(__c33__) || defined(__microblaze__)) && !defined(__UCLIBC_USE_TIME64__))
     unsigned long int __uclibc_unused3;
 #endif
     __pid_t shm_cpid;			/* pid of creator */
diff --git a/libc/sysdeps/linux/common/bits/sem.h b/libc/sysdeps/linux/common/bits/sem.h
--- a/libc/sysdeps/linux/common/bits/sem.h
+++ b/libc/sysdeps/linux/common/bits/sem.h
@@ -45,8 +45,8 @@
 #else
   __time_t sem_otime;			/* last semop() time */
 #endif
-#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__microblaze__) && !defined(__m68k__)) || \
-    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__microblaze__) || defined(__m68k__)) && !defined(__UCLIBC_USE_TIME64__))
+#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__microblaze__) && !defined(__m68k__) && !defined(__c33__)) || \
+    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__microblaze__) || defined(__m68k__) || defined(__c33__)) && !defined(__UCLIBC_USE_TIME64__))
   unsigned long int __uclibc_unused1;
 #endif
 #if defined(__UCLIBC_USE_TIME64__)
@@ -55,8 +55,8 @@
 #else
   __time_t sem_ctime;			/* last time changed by semctl() */
 #endif
-#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__microblaze__) && !defined(__m68k__)) || \
-    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__microblaze__) || defined(__m68k__)) && !defined(__UCLIBC_USE_TIME64__))
+#if (__WORDSIZE == 32 && !defined(__ARC64_ARCH32__) && !defined(__arc__) && !defined(__arm__) && !defined(__or1k__) && !defined(__xtensa__) && !defined(__riscv) && !defined(__csky__) && !defined(__i386__) && !defined(__microblaze__) && !defined(__m68k__) && !defined(__c33__)) || \
+    ((defined(__ARC64_ARCH32__) || defined(__arc__) || defined(__arm__) || defined(__or1k__) || defined(__xtensa__) || defined(__riscv) || defined(__csky__) || defined(__i386__) || defined(__microblaze__) || defined(__m68k__) || defined(__c33__)) && !defined(__UCLIBC_USE_TIME64__))
   unsigned long int __uclibc_unused2;
 #endif
   unsigned long int sem_nsems;		/* number of semaphores in set */
diff --git a/libc/misc/sysvipc/shm.c b/libc/misc/sysvipc/shm.c
--- a/libc/misc/sysvipc/shm.c
+++ b/libc/misc/sysvipc/shm.c
@@ -78,7 +78,7 @@
 #elif defined(__SHMID_DS_TIME64_SPLIT)
 	// combine the kernel's split time words
 	// When cmd is IPC_RMID, buf should be NULL.
-	if (buf != NULL) {
+	if (__ret >= 0 && buf != NULL && (cmd == IPC_STAT || cmd == SHM_STAT || cmd == 15 /* SHM_STAT_ANY */)) {
 		buf->shm_atime = (__time_t)buf->__shm_atime_internal_1 | (__time_t)(buf->__shm_atime_internal_2) << 32;
 		buf->shm_dtime = (__time_t)buf->__shm_dtime_internal_1 | (__time_t)(buf->__shm_dtime_internal_2) << 32;
 		buf->shm_ctime = (__time_t)buf->__shm_ctime_internal_1 | (__time_t)(buf->__shm_ctime_internal_2) << 32;
diff --git a/libc/misc/sysvipc/msgq.c b/libc/misc/sysvipc/msgq.c
--- a/libc/misc/sysvipc/msgq.c
+++ b/libc/misc/sysvipc/msgq.c
@@ -29,7 +29,7 @@
 #if (__WORDSIZE == 32) && defined(__UCLIBC_USE_TIME64__) && (defined(__mips) || defined(__MSQID_DS_TIME64_SPLIT))
 	union msqun arg = {.buff = buf};
 	// When cmd is IPC_RMID, buf should be NULL.
-	if (arg.__pad != NULL) {
+	if (__ret >= 0 && arg.__pad != NULL && (cmd == IPC_STAT || cmd == MSG_STAT || cmd == 13 /* MSG_STAT_ANY */)) {
 		arg.buff->msg_stime = (__time_t)arg.buff->msg_stime_internal_1 | (__time_t)(arg.buff->msg_stime_internal_2) << 32;
 		arg.buff->msg_rtime = (__time_t)arg.buff->msg_rtime_internal_1 | (__time_t)(arg.buff->msg_rtime_internal_2) << 32;
 		arg.buff->msg_ctime = (__time_t)arg.buff->msg_ctime_internal_1 | (__time_t)(arg.buff->msg_ctime_internal_2) << 32;
diff --git a/libc/misc/sysvipc/sem.c b/libc/misc/sysvipc/sem.c
--- a/libc/misc/sysvipc/sem.c
+++ b/libc/misc/sysvipc/sem.c
@@ -48,20 +48,28 @@
 
 int semctl(int semid, int semnum, int cmd, ...)
 {
-    union semun arg;
+    union semun arg = { .__pad = NULL };
     va_list ap;
 
-    /* Get the argument.  */
-    va_start (ap, cmd);
-    arg = va_arg (ap, union semun);
-    va_end (ap);
+    /* Only commands with a fourth argument may read the varargs. */
+    switch (cmd) {
+    case SETVAL: case SETALL: case GETALL:
+    case IPC_STAT: case IPC_SET: case IPC_INFO:
+    case SEM_STAT: case SEM_INFO: case 20 /* SEM_STAT_ANY */:
+        va_start (ap, cmd);
+        arg = va_arg (ap, union semun);
+        va_end (ap);
+        break;
+    default:
+        break;
+    }
 #ifdef __NR_semctl
     int __ret = __semctl(semid, semnum, cmd | __IPC_64, arg.__pad);
 #if (__WORDSIZE == 32) && defined(__UCLIBC_USE_TIME64__)
     // Only when cmd is IPC_STAT and IPC_SET, semun points to struct semid_ds.
     // At this point, arg.__pad should not be NULL, but a check is added just
     // to be safe.
-    if ((cmd & (IPC_STAT | IPC_SET)) && (arg.__pad != NULL)) {
+    if (__ret >= 0 && arg.__pad != NULL && (cmd == IPC_STAT || cmd == SEM_STAT || cmd == 20 /* SEM_STAT_ANY */)) {
         arg.buf->sem_otime = (__time_t)arg.buf->__sem_otime_internal_1 | (__time_t)(arg.buf->__sem_otime_internal_2) << 32;
         arg.buf->sem_ctime = (__time_t)arg.buf->__sem_ctime_internal_1 | (__time_t)(arg.buf->__sem_ctime_internal_2) << 32;
     }
